linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,187 @@
+# completion_port_linux requirements
+
+## Overview
+
+The completion_port_linux module handles the threading and control of the epoll socket system
+
+## Exposed API
+
+```C
+typedef struct COMPLETION_PORT_TAG* COMPLETION_PORT_HANDLE;
+
+#define COMPLETION_PORT_EPOLL_ACTION_VALUES \
+    COMPLETION_PORT_EPOLL_EPOLLRDHUP, \
+    COMPLETION_PORT_EPOLL_EPOLLIN, \
+    COMPLETION_PORT_EPOLL_EPOLLOUT, \
+    COMPLETION_PORT_CLOSING
+
+MU_DEFINE_ENUM(COMPLETION_PORT_EPOLL_ACTION, COMPLETION_PORT_EPOLL_ACTION_VALUES)
+
+#define COMPLETION_PORT_EPOLL_RESULT_VALUES \
+    COMPLETION_PORT_RESULT_OK, \
+    COMPLETION_PORT_RESULT_ERROR, \
+    COMPLETION_PORT_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(COMPLETION_PORT_RESULT, COMPLETION_PORT_RESULT_VALUES)
+
+typedef void (*ON_COMPLETION_PORT_EVENT_COMPLETE)(void* context, COMPLETION_PORT_EPOLL_ACTION epoll_action, COMPLETION_PORT_RESULT result, int32_t amount_transfered);
+
+typedef struct EPOLL_BUFFER_DATA_TAG
+{
+    void* buffer;
+    uint32_t length;
+} EPOLL_BUFFER_DATA;
+
+typedef struct COMPLETION_PORT_INGRESS_DATA_TAG
+{
+    int epoll_op;
```

Comment:
>epoll_op [](http://example.com/codeflow?start=8&length=8)

This seems to be passed to add.
I thought that we want to bind together in a handle one epoll and one thread.
So it would mean that the epoll is passed to the create me thinks.

Adding more sockets to the "completion port" would be allowed obviously.

But I think the key thing was to not have multiple threads process events from one epoll (because order is important).

Reply:
This is the epoll_operation value.  I will change the name so it makes more sense.